tls-options/client-tls-peerauth.c

```diff
@@ -0,0 +1,359 @@
+ * (at your option) any later version.
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/* the usual suspects */
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+/* socket includes */
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+/* wolfSSL */
+#include <wolfssl/options.h>
+#include <wolfssl/ssl.h>
+
+#define DEFAULT_PORT 11111
+
+enum {
+    VERIFY_OVERRIDE_ERROR,
+    VERIFY_FORCE_FAIL,
+    VERIFY_USE_PREVERIFY,
+    VERIFY_DEFAULT,
+    VERIFY_OVERRIDE_DATE_ERR
+};
+
+
+
+static THREAD_LS_T int myVerifyAction = VERIFY_DEFAULT;
+
+static WC_INLINE int myVerify(int preverify, WOLFSSL_X509_STORE_CTX* store)
+{
+    char buffer[WOLFSSL_MAX_ERROR_SZ];
+    WOLFSSL_X509* peer;
+
+
+
+    fprintf(stderr, "In verification callback, error = %d, %s\n", store->error,
+                                 wolfSSL_ERR_error_string(store->error, buffer));
+    peer = store->current_cert;
+    if (peer) {
+        char* issuer  = wolfSSL_X509_NAME_oneline(
+                                       wolfSSL_X509_get_issuer_name(peer), 0, 0);
+        char* subject = wolfSSL_X509_NAME_oneline(
+                                      wolfSSL_X509_get_subject_name(peer), 0, 0);
+        printf("\tPeer's cert info:\n issuer : %s\n subject: %s\n", issuer,
+                                                                  subject);
+
+        XFREE(subject, 0, DYNAMIC_TYPE_OPENSSL);
+        XFREE(issuer,  0, DYNAMIC_TYPE_OPENSSL);
+    }
+    else
+        fprintf(stderr, "\tPeer has no cert!\n");
+
+    printf("\tSubject's domain name at %d is %s\n", store->error_depth, store->domain);
+
+    switch (myVerifyAction) {
+        case VERIFY_DEFAULT:
+        case VERIFY_OVERRIDE_ERROR:
+            return 1;
+
+        case VERIFY_FORCE_FAIL:
+            return 0;
+
+        case VERIFY_USE_PREVERIFY:
+            return preverify;
+        
+        case VERIFY_OVERRIDE_DATE_ERR:
+            if (store->error == ASN_BEFORE_DATE_E || store->error == ASN_AFTER_DATE_E)
+                return 1;
+            return 0;
+    }
+    exit(1);
+}
+
+int switchPeerAuthMode(char *mode)
+{
+    if (XSTRCMP("NONE", mode) == 0) {
+        return SSL_VERIFY_NONE;
+    }
+    else if (XSTRCMP("PEER", mode) == 0) {
+        return SSL_VERIFY_PEER;
+    }
+    else if (XSTRCMP("FAIL_IN_NO_PEER_CERT", mode) == 0) {
+        return SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+    }
+    else if (XSTRCMP("FAIL_EXCEPT_PSK", mode) == 0) {
+        return SSL_VERIFY_FAIL_EXCEPT_PSK;
+    }
+    exit(1);
```

Comment:
When the control comes here, exit() will terminate the process leaving no message to users.  Could you leave some message or replace exit with return so that the caller can handle the error?